Fixes #16759 - ensures OSM layers use OGR random layer read/write

stev-0 opened the pull request:
Guarantee all features are rendered with large files.

## Description
Any OSM files larger than around 50MB (could be smaller) seemed to have trouble being rendered directly in QGIS via OGR. It seems that GDAL 2.2 introduced a new reading mode specifically designed for OSM files, which have a tendency to run out of memory if they are read sequentially due to the large number of nodes which must be stored in memory to create ways (linestrings). This was not being used in QGIS, so this PR just switches the new API on for OSM layers only.

More detail is available here https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite. Please see the issue on [redmine](https://issues.qgis.org/issues/16759) for pictures pre and post the PR.

Please could someone review the code here, my C++ is not yet very reliable.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I ave read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

stev-0 committed:
Fixes #16759 - ensures OSM layers use OGR random layer read/write
To guzrantee all features are rendered with large files

stev-0 committed:
Fix indentation issues

nyalldawson commented:
@rouault?

rouault commented on src/providers/ogr/qgsogrfeatureiterator.cpp:
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
-  while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
+  // OSM layers (especially large ones) need the GDALDataset::GetNextFeature() call rather than OGRLayer::GetNextFeature()
+  // see more details here: https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) ) 
```
this should be protected by #ifdef GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,2,0)

rouault commented on src/providers/ogr/qgsogrfeatureiterator.cpp:
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
-    return true;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, nullptr, nullptr, nullptr, nullptr )), fet )
+    {
+      if ( checkFeature( fet, feature ) ) 
```
this should check that the feature belongs to the layer of interest. It is returned in the second parameter of GDALDatasetGetNextFeature()

m-kuhn commented on src/providers/ogr/qgsogrfeatureiterator.cpp:
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
+      {
+        return true;
+      }
+      else 
```
Is the else branch here required?

stev-0 committed:
Respond to review of 16759
Add check for GDAL 2.3, check correct OGR layer, removed unecessary code paths and better formatting

nyalldawson commented on src/providers/ogr/qgsogrfeatureiterator.cpp:
```diff
@@ -282,20 +296,41 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
+    OGRLayerH nextFeatureBelongingLayer;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, &nextFeatureBelongingLayer, nullptr, nullptr, nullptr ) ), fet )
+    {
+      if ( checkFeature( fet, feature ) && QString::compare( QString::fromLatin1( OGR_L_GetName( nextFeatureBelongingLayer ) ), mSource->mLayerName ) == 0 )
```
I think this check should be reversed -- the layer name comparison is much cheaper than the checkFeature, which requires reading the feature itself.

rouault replied:
And you can replace the layer name comparison by checking the equality of layer pointers directly.

rouault commented on src/providers/ogr/qgsogrfeatureiterator.cpp:
```diff
@@ -305,7 +305,7 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
     while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, &nextFeatureBelongingLayer, nullptr, nullptr, nullptr ) ), fet )
     {
-      if ( checkFeature( fet, feature ) && QString::compare( QString::fromLatin1( OGR_L_GetName( nextFeatureBelongingLayer ) ), mSource->mLayerName ) == 0 )
+      if ( nextFeatureBelongingLayer == mSource->mLayerName && checkFeature( fet, feature ) )
```
huh ! does this actually compile ??? You're comparing a OGRLayer* pointer to a QString !

stev-0 replied:
My bad - not at my workstation, so shortcutting. That'll teach me

stev-0 committed:
Simplify OGR layer check and make quicker check first

stev-0 commented:
This should be all done now. Let me know if I should squash commits

nyalldawson commented:
Thanks @stev-0!
